src/java.desktop/share/classes/java/awt/geom/Path2D.java

```diff
@@ -2129,6 +2081,140 @@ public final Rectangle getBounds() {
+     * </p>
+     * @return an instance of {@code Rectangle2D} that is a high-precision bounding box of the
+     *         {@code PathIterator}.
+     * @see Shape#getBounds2D()
+     */
+    public static Rectangle2D getBounds2D(PathIterator pi) {
+        // define x and y parametric coefficients where:
+        // x(t) = x_coeff[0] + x_coeff[1] * t + x_coeff[2] * t^2 + x_coeff[3] * t^3
+        double[] x_coeff = new double[4];
+        double[] y_coeff = new double[4];
+
+        double[] coords = new double[6];
+        double[] tExtrema = new double[3];
+        boolean isEmpty = true;
+        double leftX = 0.0;
+        double rightX = 0.0;
+        double topY = 0.0;
+        double bottomY = 0.0;
+        double lastX = 0.0;
+        double lastY = 0.0;
+
+        pathIteratorLoop : while (!pi.isDone()) {
+            int type = pi.currentSegment(coords);
+            pi.next();
+            double endX, endY;
+            switch (type) {
+                case PathIterator.SEG_MOVETO, PathIterator.SEG_LINETO:
+                    endX = coords[0];
+                    endY = coords[1];
+                    break;
+                case PathIterator.SEG_QUADTO:
+                    endX = coords[2];
+                    endY = coords[3];
+                    break;
+                case PathIterator.SEG_CUBICTO:
+                    endX = coords[4];
+                    endY = coords[5];
+                    break;
+                default:
+                    continue pathIteratorLoop;
+            }
+
+            if (isEmpty) {
+                // we're seeding our bounds for the first time:
+                isEmpty = false;
+                leftX = rightX = endX;
+                topY = bottomY = endY;
+            } else {
+                // extend our rectangle to cover the point at t = 1:
+                leftX = (endX < leftX) ? endX : leftX;
+                rightX = (endX > rightX) ? endX : rightX;
+                topY = (endY < topY) ? endY : topY;
+                bottomY = (endY > bottomY) ? endY : bottomY;
+            }
+
+            // here's the slightly trickier part: examine quadratic and cubic
+            // segments for extrema where t is between (0, 1):
+
+            boolean definedParametricEquations;
+            if (type == PathIterator.SEG_QUADTO) {
+                definedParametricEquations = true;
+
+                x_coeff[3] = 0.0;
+                x_coeff[2] = lastX - 2.0 * coords[0] + coords[2];
+                x_coeff[1] = -2.0 * lastX + 2.0 * coords[0];
+                x_coeff[0] = lastX;
+
+                y_coeff[3] = 0;
+                y_coeff[2] = lastY - 2.0 * coords[1] + coords[3];
+                y_coeff[1] = -2.0 * lastY + 2.0 * coords[1];
+                y_coeff[0] = lastY;
+            } else if (type == PathIterator.SEG_CUBICTO) {
+                definedParametricEquations = true;
+
+                x_coeff[3] = -lastX + 3.0 * coords[0] - 3.0 * coords[2] + coords[4];
```

Comment:
To improve coefficient accuracy ~ few ulps, it should be written explicitely as differences (= vector notation).
See Marlin DCurve formula:
`    void set(final double x1, final double y1,
             final double x2, final double y2,
             final double x3, final double y3,
             final double x4, final double y4)
    {
        final double dx32 = 3.0 * (x3 - x2);
        final double dx21 = 3.0 * (x2 - x1);
        ax = (x4 - x1) - dx32;  // A = P3 - P0 - 3 (P2 - P1) = (P3 - P0) + 3 (P1 - P2)
        bx = (dx32 - dx21);     // B = 3 (P2 - P1) - 3(P1 - P0) = 3 (P2 + P0) - 6 P1
        cx = dx21;              // C = 3 (P1 - P0)
        dx = x1;                // D = P0
        dax = 3.0 * ax;
        dbx = 2.0 * bx;

        final double dy32 = 3.0 * (y3 - y2);
        final double dy21 = 3.0 * (y2 - y1);
        ay = (y4 - y1) - dy32;
        by = (dy32 - dy21);
        cy = dy21;
        dy = y1;
        day = 3.0 * ay;
        dby = 2.0 * by;
    }

    void set(final double x1, final double y1,
             final double x2, final double y2,
             final double x3, final double y3)
    {
        final double dx21 = (x2 - x1);
        ax = 0.0;              // A = 0
        bx = (x3 - x2) - dx21;  // B = P3 - P0 - 2 P2
        cx = 2.0 * dx21;       // C = 2 (P2 - P1)
        dx = x1;                // D = P1
        dax = 0.0;
        dbx = 2.0 * bx;

        final double dy21 = (y2 - y1);
        ay = 0.0;
        by = (y3 - y2) - dy21;
        cy = 2.0 * dy21;
        dy = y1;
        day = 0.0;
        dby = 2.0 * by;
    }`